File: android/app/src/main/java/com/nxtend/team35/yubiboard/vision/TrackingStateMachine.kt
package com.nxtend.team35.yubiboard.vision

enum class TrackingState {
    UNDETECTED,
    CANDIDATE,
    TRACKING,
    TEMPORARILY_LOST,
}

class TrackingStateMachine(
    private val detectionsRequired: Int = 3,
    private val lossTimeoutMs: Long = 300,
) {
    var state: TrackingState = TrackingState.UNDETECTED
        private set
    private var consecutiveDetections = 0
    private var lostAtMs: Long? = null

    fun update(detected: Boolean, timestampMs: Long): TrackingState {
        if (detected) {
            lostAtMs = null
            consecutiveDetections++
            state = when (state) {
                TrackingState.TRACKING,
                TrackingState.TEMPORARILY_LOST,
                -> TrackingState.TRACKING
                TrackingState.UNDETECTED,
                TrackingState.CANDIDATE,
                -> if (consecutiveDetections >= detectionsRequired) {
                    TrackingState.TRACKING
                } else {
                    TrackingState.CANDIDATE
                }
            }
        } else {
            consecutiveDetections = 0
            state = when (state) {
                TrackingState.TRACKING -> {
                    lostAtMs = timestampMs
                    TrackingState.TEMPORARILY_LOST
                }
                TrackingState.TEMPORARILY_LOST -> {
                    if (timestampMs - (lostAtMs ?: timestampMs) >= lossTimeoutMs) {
                        lostAtMs = null
                        TrackingState.UNDETECTED
                    } else {
                        TrackingState.TEMPORARILY_LOST
                    }
                }
                else -> TrackingState.UNDETECTED
            }
        }
        return state
    }
}

File: ios/Screact/Vision/TrackingStateMachine.swift
import Foundation

enum TrackingState {
    case undetected
    case candidate
    case tracking
    case temporarilyLost
}

/// Direct port of Android's TrackingStateMachine. Promotes to `.tracking` after
/// `detectionsRequired` consecutive detections and tolerates short losses up to `lossTimeoutMs`.
final class TrackingStateMachine {
    private let detectionsRequired: Int
    private let lossTimeoutMs: Int64
    private(set) var state: TrackingState = .undetected
    private var consecutiveDetections = 0
    private var lostAtMs: Int64?

    init(detectionsRequired: Int = 3, lossTimeoutMs: Int64 = 300) {
        self.detectionsRequired = detectionsRequired
        self.lossTimeoutMs = lossTimeoutMs
    }

    @discardableResult
    func update(detected: Bool, timestampMs: Int64) -> TrackingState {
        if detected {
            lostAtMs = nil
            consecutiveDetections += 1
            switch state {
            case .tracking, .temporarilyLost:
                state = .tracking
            case .undetected, .candidate:
                state = consecutiveDetections >= detectionsRequired ? .tracking : .candidate
            }
        } else {
            consecutiveDetections = 0
            switch state {
            case .tracking:
                lostAtMs = timestampMs
                state = .temporarilyLost
            case .temporarilyLost:
                if timestampMs - (lostAtMs ?? timestampMs) >= lossTimeoutMs {
                    lostAtMs = nil
                    state = .undetected
                } else {
                    state = .temporarilyLost
                }
            default:
                state = .undetected
            }
        }
        return state
    }
}
